File: April Bootcamp/matrix_Multiplication.py
# CODE for matrix multiplication

# Function for matrix multiplication

def matrix_multiplication(A, B):
    #check if the number of columns in A is equal to the number of rows in B
    if (len(A[0]) != len(B)):
        print("Multiplication is not possible")
        return
    
    C = [[0 for _ in range(len(A))] for _ in range(len(B[0]))]
    for i in range(len(A)):
        for j in range(len(B[0])):
            for k in range(len(B)):
                C[i][j] += A[i][k] * B[k][j]
    
    for i in range(len(C)):
        print(C[i])



#__MAIN__
A = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
B = [[9, 8, 7], [6, 5, 4], [3, 2, 1]]

# A list D for storing elements after mltiplication of A and B using list comprehension
# D = [sum(a*b for a, b in zip(A_row, B_col)) for A_row in A for B_col in zip(*B)]

matrix_multiplication(A, B)

File: C++/2D Arrays/matrix_multiplication.cpp
#include<iostream>
using namespace std;

int main(){

    int n, m;
    cout<<"Enter number of rows and columns for matix A: ";
    cin>>m>>n;
    int a[m][n];
    // taking input for matrix A
    cout<<"Enter elements : \n";
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            cin>>a[i][j];
        }
    }

    cout<<"===================================\n";

    int p, q;
    cout<<"Enter number of rows and columns for matrix B: ";
    cin>>p>>q;
    int b[p][q]; 
    // taking input for matrix B
    cout<<"Enter elements : \n";
    for (int i = 0; i < p; i++) {
        for (int j = 0; j < q; j++) {
            cin>>b[i][j];
        }
    } 

    cout<<"===================================\n";

    if(n != p) cout << "Matrix Multiplication not possible !!!";

    int c[m][q];  // matrix for storing product of A and B
    for(int i=0; i<m;i++){
        for(int j=0; j<q; j++){
            c[i][j] = 0;
        }
    }

    int rowA = 0;
    while(rowA < m){
        int colB = 0;
        for(int i=0; i<q; i++){  // q is the number of columns of B
            int j=0;
            while(j<p){  // n and p are equal
                c[rowA][colB] += a[rowA][j] * b[j][colB];
                j++;
            }
            colB++;
        }
        rowA++;
    }

    // printing the product matrix;
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < q; j++) {
            cout<<c[i][j]<<" "; 
        }
        cout<<endl;
    }

}